src/cpp/session/modules/SessionCodeSearch.cpp

```diff
@@ -1543,6 +1552,100 @@ SourceItem fromCppDefinition(const clang::CppDefinition& cppDefinition)
       safe_convert::numberTo<int>(cppDefinition.location.column, 1));
 }
 
+void fillFromBookdownRefs(const std::string& term,
+                          std::vector<SourceItem>* pSourceItems)
+{
+   // retrieve refs for this project
+   core::json::Value bookdownIndex = module_context::bookdownXRefIndex();
+   core::debug::print(bookdownIndex);
+   
+   // may be null if we have no bookdown refs (typically implies
+   // we're not in a bookdown project)
+   if (!bookdownIndex.isObject())
+      return;
+   
+   std::string baseDir;
+   core::json::Array bookdownRefs;
+   
+   Error error = core::json::readObject(
+            bookdownIndex.getObject(),
+            "baseDir", baseDir,
+            "refs", bookdownRefs);
+   
+   if (error)
+   {
+      LOG_ERROR(error);
+      return;
+   }
+   
+   FilePath basePath = module_context::resolveAliasedPath(baseDir);
+   
+   // {
+   //     "file": "01-intro.Rmd",
+   //     "type": "h1",
+   //     "id": "intro",
+   //     "title": "Introduction"
+   // }
+   
+   for (const json::Value& bookdownRef : bookdownRefs)
+   {
+      if (!bookdownRef.isObject())
+         continue;
+      
+      std::string file, type, id, title;
+      Error error = core::json::readObject(
+               bookdownRef.getObject(),
+               "file", file,
+               "type", type,
+               "id", id,
+               "title", title);
+      
+      if (error)
+      {
+         LOG_ERROR(error);
+         continue;
+      }
+      
+      // figure out appropriate source item type
+      SourceItem::Type sourceType = SourceItem::None;
```

Comment:
@jjallaire: are there any other reference types I should handle here?

Reply:
All the types are here (save for `eq` which I appear to have forgotten!): https://github.com/rstudio/rstudio/blob/master/src/gwt/panmirror/src/editor/src/api/xref.ts#L99